### DISCUSSION TITLE
InteriorFaceIntegrator for trial != test space

### DISCUSSION THREAD

**Discussion by alecampos2009** [mfem/mfem#3176] [General]
Dear MFEM developers, community,

I have a PDE for which I use the standard diffusion integrator for one of its terms. However, this specific term in the PDE has a trial space that is different from the test space (the trial space is H1 and the test space is L2).

My understanding is that, along with this diffusion integrator, a `DGDiffusionIntegrator()` needs to be added using the `AddInteriorFaceIntegrator()`. However, since the trial and test spaces are different, I get a run-time MFEM abort message stating trial != test space is not supported.

Any suggestions on an appropriate method to overcome this would be much appreciated. Is the suggested fix to project the trial space from H1 to L2?

Thank you,
Alejandro

**Comment by vladotomov** [comment]
Hi @alecampos2009,

Sorry, we must have missed this question.

The mfem's implementation of `DGDiffusionIntegrator` assumes the same L2 space for test and trial. You can inherit from `BilinearFormIntegrator` and implement a mixed version by defining the method
```
   /** Abstract method used for assembling TraceFaceIntegrators in a
       MixedBilinearForm. */
   virtual void AssembleFaceMatrix(const FiniteElement &trial_face_fe,
                                   const FiniteElement &test_fe1,
                                   const FiniteElement &test_fe2,
                                   FaceElementTransformations &Trans,
                                   DenseMatrix &elmat);
```
and then add the new integrator as a `TraceFaceIntegrator` in a `MixedBilinearForm`.

What have you tried and what exactly was the error you got? Which was the function that produced it?

Best,
Vladimir
